Sync up6 branch with master (#7040)
ballerina-platform/ballerina-dev-website: swan-lake/learn-the-platform/ballerina-persist/overview.md

---
layout: ballerina-persist-documentation-left-nav-pages-swanlake
title: Bal persist overview
description: The section gives an overview of the `bal persist` feature.
keywords: ballerina, programming language, ballerina packages, persist, persist model, persist cli, persist client api

> permalink: /learn/ballerina-persist/persist-overview/

Author: ```suggestion
permalink: /learn/ballerina-persist/persist-overview/
```
Currently, the link on the Learn landing page for this page is broken. Also, the text is added as `Overview` and it should be changed to `Persist overview`.

active: persist_overview
intro: The `bal persist` feature allows you to store data in different data stores and retrieve them when needed. A data store can be a database, an in-memory cache, or a file system. The bal persist feature currently supports In-memory table, MySQL database, and Google Sheets as data stores. The important point is you can use the same syntax to access data in all these data stores. So you don't need to learn different syntaxes to access data in different data stores.

> redirect_from:

Author: ```suggestion
redirect_from:
```
Why do we need redirections as these are new content?

- /learn/ballerina-persist/persist-overview/
---

> This feature has three main sections: Persist Model, Persist CLI and Persist Client API. The Persist Model is used to define the data model. The Persist CLI is used to generate the client API for the data model. The Persist Client API is used to access the data in the data store.

Author: ```suggestion
This feature has three main components: Persist Model, Persist CLI, and Persist Client API. The Persist Model is used to define the data model. The Persist CLI is used to generate the client API for the data model. The Persist Client API is used to access the data in the data store.
```
Shall we add the 3 components as a list along with the descriptions? These descriptions might be redundant as we have sections below with all 3 described in detail.

## Persist Model

> The Persist Model is used to define the data model. The data model is defined in a Ballerina source file inside the persist directory in the project root directory. The data model is defined using the Ballerina record type. The following example shows how to define a data model for a table named `Employee` with the fields `id`, `name`, `age` and `salary`.

Author: ```suggestion
The Persist Model is used to define the data model. The data model is defined in a Ballerina source file inside the `persist` directory in the project root directory. The data model is defined using the Ballerina record type. The following example shows how to define a data model for a table named `Employee` with the fields `id`, `name`, `age`, and `salary`.
```

```ballerina
type Employee record {
    readonly int id;
    string name;
    int age;
    float salary;
}
```

> The `readonly` keyword is used to define a field as a primary key. You need to have at least one primary key field in the Entity record. The primary key field is used to uniquely identify a record in the table. The primary key field is used to generate the `get`, `update`, and `delete` operations in the client API. You can define multiple primary key fields in the Entity record.

Author: ```suggestion
The `readonly` keyword is used to define a field as a primary key. You need to have at least one primary key field in the `Entity` record. The primary key field is used to uniquely identify a record in the table. The primary key field is used to generate the `get`, `update`, and `delete` operations in the client API. You can define multiple primary key fields in the `Entity` record.
```

> Learn about how to define relationships between entity records and more about the persist model in the [Persist Model](/learn/persist-model/) section.

Author: ```suggestion
Learn about how to define relationships between entity records and more about the persist model in the [Persist Model](/learn/persist-model/) section.
```
Refer the guidelines on how to add hyperlinks.

> > Note: The Ballerina VSCode plugin provides validation and code actions for the persist model. So you can easily create the persist model using the Ballerina VSCode plugin.

Author: ```suggestion
> **Note:** The Ballerina VS Code plugin provides validation and code actions for the persist model. Therefore, you can easily create the Persist Model using the Ballerina VS Code plugin.
```

## Persist CLI

> The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable file and setup scripts are generated. For example, if you are using MySQL as the data store, the `persist_db_config.bal` file, and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the MySQL database connection. The `script.sql` file is used to create the table in the MySQL database. The client API is generated in the `generated` directory in the project root directory.

Author: ```suggestion
The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable files and setup scripts are generated. For example, if you are using MySQL as the data store, the `persist_db_config.bal` file and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the MySQL database connection. The `script.sql` file is used to create the table in the MySQL database. The client API is generated in the `generated` directory in the project root directory.
```
This paragraph is too lengthy. Shall we consider breaking them into multiple or summarizing them? Else, we can add a list.

> The `bal persist` has two main built-in CLI commands: `persist init` and `persist generate`. The `persist init` command is used to initialize bal persist in the Ballerina project. The `persist generate` command is used to generate the client API for the data model. Additionally, there is an experimental `persist migrate` command to generate the SQL scripts for changing the table structure in the database when the data model is changed.

Author: ```suggestion
The `bal persist` has two main built-in CLI commands: `persist init` and `persist generate`. The `persist init` command is used to initialize `bal persist` in the Ballerina project. The `persist generate` command is used to generate the client API for the data model. Additionally, there is an experimental `persist migrate` command to generate the SQL scripts for changing the table structure in the database when the data model is changed.
```
This paragraph is too lengthy. Shall we consider breaking them into multiple or summarizing them? Else, we can add a list.

> Learn more about `bal persist` CLI in the [Persist CLI](/learn/persist-cli/) section.

Author: ```suggestion
Learn more about `bal persist` CLI in the [Persist CLI](/learn/persist-cli/) section.
```
Refer the guidelines on adding hyperlinks.

## Persist Client API

> The Persist Client API is used to access the data in the data store. The client API is generated using the Persist CLI and it is located in the `generated` directory in the project root directory. The client API consists of a Ballerina client object with resource for each entity record modeled in the Ballerina project. For each client resource, there are resource methods to perform CRUD operations on the data store.

Author: ```suggestion
The Persist Client API is used to access the data in the data store. The client API is generated using the Persist CLI and it is located in the `generated` directory in the project root directory. The client API consists of a Ballerina client object with a resource for each entity record modeled in the Ballerina project. For each client resource, there are resource methods to perform CRUD operations on the data store.
```

> For example, if you have an `Employee` entity record in the Ballerina project, you can perform CRUD operations on the `Employee` table in the data store using the client object, as follows.

Author: ```suggestion
For example, if you have an `Employee` entity record in the Ballerina project, you can perform CRUD operations on the `Employee` table in the data store using the client object as follows.
```

```ballerina

> // Create a new employee record

Author: ```suggestion
// Create a new `employee` record.
```

EmployeeInsert employee = {id: 1, name: "John", age: 30, salary: 3000.0};
int[]|error employeeId = sClient->/employees.post([employee]);

> // Get the employee record with id 1

Author: ```suggestion
// Get the `employee` record with the ID 1.
```

Employee|error employee = sClient->/employees/1;

> // Update the employee record with id 1

Author: ```suggestion
// Update the `employee` record with the ID 1.
```

Employee|error updated = sClient->/employees/1.put({salary: 4000.0});

> // Delete the employee record with id 1

Author: ```suggestion
// Delete the employee record with the ID 1.
```

Employee|error deleted = sClient->/employees/1.delete();

> // Get all the employee records

Author: ```suggestion
// Get records of all employees.
```

stream<Employee, error?> employees = sClient->/employees;
```

> Learn more about persist client APIs in the [Persist Client API](/learn/persist-client-api/) section.

Author: ```suggestion
Learn more about persist client APIs in the [Persist Client API](/learn/persist-client-api/) section.
```
Refer the guidelines on adding hyperlinks.

> If you want to get started with a practical introduction and learn about `bal persist`, head over to the [Quick Start Guide](quick-tour.md) section.

Author: ```suggestion
If you want to get started with a practical introduction and learn about `bal persist`, head over to the [Quick Start Guide](quick-tour.md) section.
```
Refer the guidelines on adding hyperlinks. Avoid marketing jargon.
